packages/ui/src/components/Menu/SMenuItem.vue

```diff
@@ -0,0 +1,71 @@
+<script lang="ts">
+export default { name: 'SMenuItem' }
+</script>
+
+<script setup lang="ts">
+import { computed, onUnmounted } from 'vue'
+import { useMenuApi, useSubmenuApi } from '@/components/Menu/api'
+import SMenuItemBody from '@/components/Menu/SMenuItemBody.vue'
+
+const props = withDefaults(
+  defineProps<{
+    index: string
+  }>(),
+  {},
+)
+
+const emit = defineEmits<(event: 'click', value: HTMLElement) => void>()
+
+let isSelected = ref(false)
+let isMenuCollapsed = ref(false)
+const menuApi = useMenuApi()
+
+if (menuApi) {
+  isSelected = computed(() => menuApi.active.value === props.index)
+  isMenuCollapsed = computed(() => menuApi.collapsed.value)
+}
+
+function handleClick(event: HTMLElement) {
+  if (menuApi) {
+    menuApi.select(props.index)
+  }
+
+  emit('click', event)
+}
+
+const isInSubmenu = ref(false)
+const submenuApi = useSubmenuApi()
+
+if (submenuApi) {
+  isInSubmenu.value = true
+
+  submenuApi.register(props.index)
+
+  onUnmounted(() => {
+    submenuApi.unregister(props.index)
+  })
```

Comment:
This mechanism seems fragile - what if there are multiple menu items with the same index? Also, what if `index` is reactive?

Reply:
My idea:

```ts
import { InjectionKey, Ref, computed, inject, onScopeDispose, provide, ref, shallowReactive } from 'vue'

interface Api {
  register: (value: Ref<string>) => { isActive: Ref<boolean>; select: () => void }
}

const API_KEY: InjectionKey<Api> = Symbol('something')

// in parent

function setupParent(props: { modelValue: string }) {
  const activeValue = computed({
    get: () => props.modelValue,
    set: (v) => {
      // updating model value
    },
  })

  const registered = shallowReactive<Set<{ value: Ref<string> }>>(new Set())

  provide(API_KEY, {
    register: (value) => {
      const obj = { value }
      registered.add(obj)

      onScopeDispose(() => registered.delete(obj))

      return {
        isActive: computed(() => activeValue.value === value.value),
        select: () => {
          activeValue.value = value.value
        },
      }
    },
  })
}

// in child

function setupChild(props: { value: string }) {
  const value = computed(() => props.value)

  const { register } = inject(API_KEY)!

  const { isActive, select } = register(value)

  return { isActive, select }
}

```